crates/kitsune_p2p/types/src/dht_arc.rs

```diff
@@ -20,6 +24,111 @@ pub struct DhtLocation(pub Wrapping<u32>);
+/// as a difference of 1.0 would scale to 0.2.
+const DELTA_SCALE: f64 = 0.2;
+
+/// The minimal "change in arc" before we stop scaling.
+/// This prevents never reaching the target arc coverage.
+const DELTA_THRESHOLD: f64 = 0.01;
+
+/// Due to estimation noise we don't want a very small difference
+/// between observed coverage and estimated coverage to
+/// amplify when scaled to by the estimated total peers.
+/// This threshold must be reached before an estimated coverage gap
+/// is calculated.
+const NOISE_THRESHOLD: f64 = 0.01;
+
+#[cfg(any(test, feature = "test_utils"))]
+/// Margin of error for floating point comparisons
+const ERROR_MARGIN: f64 = 0.0000000001;
+
+// TODO: Use the [`f64::clamp`] when we switch to rustc 1.50
+fn clamp(min: f64, max: f64, mut x: f64) -> f64 {
+    if x < min {
+        x = min;
+    }
+    if x > max {
+        x = max;
+    }
+    x
+}
+
+/// The ideal coverage if all peers were holding the same sized
+/// arcs and our estimated total peers is close.
+fn coverage_target(est_total_peers: usize) -> f64 {
+    if est_total_peers <= REDUNDANCY_TARGET {
+        1.0
+    } else {
+        REDUNDANCY_TARGET as f64 / est_total_peers as f64
+    }
+}
+
+/// Calculate the target arc length given a peer density.
+fn target(density: PeerDensity) -> f64 {
+    // Get the estimated coverage gap based on our observed peer density.
+    let est_gap = density.est_gap();
+    // If we haven't observed at least our redundancy target number
+    // of peers (adjusted for expected uptime) then we know that the data
+    // in our arc is under replicated and we should start aiming for full coverage.
+    if density.expected_count() < REDUNDANCY_TARGET {
+        1.0
+    } else {
+        // Get the estimated gap. We don't care about negative gaps
+        // or gaps we can't fill (> 1.0)
+        let est_gap = clamp(0.0, 1.0, est_gap);
+        // Get the ideal coverage target for the size of that we estimate
+        // the network to be.
+        let ideal_target = coverage_target(density.est_total_peers());
+        // Take whichever is larger. We prefer nodes to target the ideal
+        // coverage but if there is a larger gap then it needs to be filled.
+        let target = est_gap.max(ideal_target);
+
+        clamp(0.0, 1.0, target)
+    }
+}
+
+/// The convergence algorithm that moves an arc towards
+/// our estimated target.
+///
+/// Note the rate of convergence is dependant of the rate
+/// that [`DhtArc::update_length`] is called.
+fn converge(current: f64, density: PeerDensity) -> f64 {
+    let target = target(density);
+    // The change in arc we'd need to make to get to the target.
+    let delta = target - current;
+    // If this is below our threshold then apply that delta.
+    if delta.abs() < DELTA_THRESHOLD {
+        current + delta
+    // Other wise scale the delta to avoid rapid change.
+    } else {
+        current + (delta * DELTA_SCALE)
+    }
```

Comment:
This seems weird to have this sharp discontinuity. As soon as the delta goes over the threshold, the effective delta becomes 20% of what it would have been if it were only a little less.

Seems like since you expect to call `update_length` often, and concurrently with updates to other arcs, you'd want each call to nudge the length closer to the target, but never completely. The delta scaling could be adaptive, but seems like it should be smooth.

Reply:
Yeh the idea is that it's so small we don't care but yeh it could be better.
Also I'm not sure how often `update_length` will be called but currently our agent info lasts 20 minutes so it could be that slow.